framework/src/problems/FEProblemBase.C

```diff
@@ -3332,64 +3328,86 @@ FEProblemBase::projectSolution()
   FloatingPointExceptionGuard fpe_guard(_app);
 
   ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
-  ComputeInitialConditionThread cic(*this);
-  Threads::parallel_reduce(elem_range, cic);
 
-  if (haveFV())
+  // loop over _global_current_state = 2, 1, 0
+  for (_global_current_state = 2; _global_current_state >= 0; _global_current_state--)
   {
-    using ElemInfoRange = StoredRange<MooseMesh::const_elem_info_iterator, const ElemInfo *>;
-    ElemInfoRange elem_info_range(_mesh.ownedElemInfoBegin(), _mesh.ownedElemInfoEnd());
+    ComputeInitialConditionThread cic(*this);
+    Threads::parallel_reduce(elem_range, cic);
```

Comment:
Yes. This was required to avoid more complicated logic that would be prone to more edge cases. Here are the design requirements:

1. _global_current_state has to exist inside FEProblemBase. This was to respect the ordering of IC computing that happens in FEProblemBase::projectSolution(). In that function, someone decided that there was a particular order that ICs needed to be computed in. However, inside ComputeInitialConditionThread::compute() was the approperiate place to apply the logic of which ICs to compute. 

2. And then, sys->copySolutionsBackwards() needs to be called from FEProblemBase, as far as I can tell. 

This led me to need some way of transferring loop_state information back and forth between FEProblemBase::projectSolution() and ComputeInitialConditionThread::compute().

3. The final requirement was that if someone only specifies initial conditions for the current state, that IC needs to be copied backward to the old and older states. This was the old functionality of MOOSE.
4. Otherwise, if someone specified multiple states, then the proper ICs needed to be copied back to their appropriate states. 

So if there is a better way to accomplish these things, I'm open to suggestions.

The way the loop inside ComputeInitialConditionThread::compute() works is that it checks to see if state=2 was specified for the variable, then state=1, then state=0. If state=2 was specified, then it will abort the loop, and the state=2 initial condition will get copied back to the appropriate state. But if someone only specified state=0 for that state, then the loop won't find any initial conditions for states 1 and 2, so it will choose the state=0 IC, and that will get copied back to the old states.

Reply:
1 & 2: sure, I didn't complain about the way you transfer state between FEProblem and ICs...

3 & 4 should definitely be documented somewhere. I personally prefer that to be part of the IC system documentation.

> The way the loop inside ComputeInitialConditionThread::compute() works is that it checks to see if state=2 was specified for the variable, then state=1, then state=0. If state=2 was specified, then it will abort the loop, and the state=2 initial condition will get copied back to the appropriate state. But if someone only specified state=0 for that state, then the loop won't find any initial conditions for states 1 and 2, so it will choose the state=0 IC, and that will get copied back to the old states.

I think I understand what you want to achieve now. I'd like to suggest a slightly different implementation because your current implementation computes the IC three times if only state=0 is set for a variable, once for state=0, once for state=1, and once for state=2. If the IC computation is very expensive, you have to evaluate the expensive IC three times!

How about this? Let's first compute IC like we used to, that is, compute IC at state=0, copy the solution vector backwards to the old solution vector (**not the older!**). Then, go to state=1, compute ICs with state=1, and copy the old solution vector to the older solution vector. Finally, go to state=2, compute ICs with state=2. This way you still preserve the old behavior when only state=0 is set for an IC, and you can avoid excessive IC evaluations by simply copying the solution vectors backwards.

The challenge with this new approach is that you need to let the threaded IC loop directly write into old and older solution vectors, instead of what you are currently doing (i.e., always writing to the current solution vector, and copy it backwards).

Reply:
Great! Actually that would be better, but in my first attempt at implementing this, I was unable to the write the old and older state solution vectors directly. Because they seem to live inside LibMesh or PetsC. Do you have any templates that I could look at without breaking a lot of Moose things? Last time there seemed to be no way to make a new vector before the old state vector was created.

Reply:
I'm not terribly familiar with that part. Could you perhaps create the old and older vectors earlier?

@dschwen any pointers?

Reply:
I worked with him on this, and he seemed to think copying backwards seemed to be the best method. But yes @dschwen please weigh in when you get back from vacation Thursday. I'll attempt to implement this before then. Thank you for the advice @hugary1995

Reply:
@dschwen and I looked at it together, and there are several layers of APIs to go through to accomplish the "writing to old solution states." I will instead implement some logic to only compute each IC object once.